src/compiler/checker.ts

```diff
@@ -18267,12 +18268,16 @@ namespace ts {
          * variable T[P] (i.e. we treat the type T[P] as the type variable we're inferring for).
          */
         function inferTypeForHomomorphicMappedType(source: Type, target: MappedType, constraint: IndexType): Type | undefined {
+            if (inInferTypeForHomomorphicMappedType) {
+                return undefined;

```

Comment:
Do we need to cache this inference failure, as in `invokeOnce`, so if it's attempted again on the same types outside the same recursive context (but during the same traversal) we still refuse to perform the secondary inference?

Reply:
I don't think so. In the general case this function is never recursive. The only code paths that could recur are special cases for array and tuple types, so we're talking something that happens very rarely. I think this simple solution is fine.